sql/src/main/java/org/elasticsearch/action/admin/indices/create/TransportBulkCreateIndicesAction.java

```diff
@@ -0,0 +1,175 @@
+ *
+ * this action enqueues requests
+ */
+public class TransportBulkCreateIndicesAction extends TransportMasterNodeOperationAction<BulkCreateIndicesRequest, BulkCreateIndicesResponse> {
+
+    public static final int QUEUE_SIZE = 100;
+    private final Set<String> queuedIndices;
+    private final BlockingQueue<RequestItem> requestQueue;
+    private final QueuedRequestExecutor queuedRequestExecutor;
+    private final MetaDataCreateIndexService createIndexService;
+
+    @Inject
+    protected TransportBulkCreateIndicesAction(Settings settings,
+                                               TransportService transportService,
+                                               ClusterService clusterService,
+                                               ThreadPool threadPool,
+                                               MetaDataCreateIndexService createIndexService,
+                                               ActionFilters actionFilters) {
+        super(settings, BulkCreateIndicesAction.NAME, transportService, clusterService, threadPool, actionFilters);
+        this.queuedIndices = Collections.synchronizedSet(new HashSet<String>());
+        this.requestQueue = new ArrayBlockingQueue<>(QUEUE_SIZE);
+        this.createIndexService = createIndexService;
+
+        this.queuedRequestExecutor = new QueuedRequestExecutor();
+        // start queued Request executor
+        this.threadPool.executor(executor()).execute(queuedRequestExecutor);
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.MANAGEMENT;
+    }
+
+    @Override
+    protected BulkCreateIndicesRequest newRequest() {
+        return new BulkCreateIndicesRequest();
+    }
+
+    @Override
+    protected BulkCreateIndicesResponse newResponse() {
+        return new BulkCreateIndicesResponse();
+    }
+
+    @Override
+    protected void masterOperation(final BulkCreateIndicesRequest request, ClusterState state, final ActionListener<BulkCreateIndicesResponse> listener) throws ElasticsearchException {
+        final List<CreateIndexResponse> responses = new ArrayList<>(request.requests().size());
+        final BulkCreateIndicesResponse finalResponse = new BulkCreateIndicesResponse(responses);
+
+        List<PlainActionFuture<ClusterStateUpdateResponse>> futures = new ArrayList<>(request.requests().size());
+        for (CreateIndexRequest createIndexRequest : request.requests()) {
+            if (!queuedIndices.contains(createIndexRequest.index())) {
+                queuedIndices.add(createIndexRequest.index());
+                try {
+                    RequestItem item = new RequestItem(createIndexRequest);
+                    futures.add(item.future);
+                    requestQueue.put(item);
+                } catch (InterruptedException e) {
+                    logger.debug("got interrupted while adding create index request to the queue");
+                    Thread.interrupted();
+                    listener.onFailure(e);
```

Comment:
either the loop should stop or the listener.onFailure call should probably be done outside the loop to avoid triggering the listener multiple times